File: ZhuZhiLiao/Physics/OrbitCounter.swift
import simd

struct OrbitCounter: Sendable {
    private var accumulatedAngle: Float = 0
    private var stableAxis: SIMD3<Float>?

    mutating func reset() {
        accumulatedAngle = 0
        stableAxis = nil
    }

    mutating func update(
        angleDelta: Float,
        axis: SIMD3<Float>,
        isQualified: Bool
    ) -> Int {
        let axisMagnitude = simd_length(axis)
        guard isQualified, angleDelta > 0, axisMagnitude > 0.000_001 else {
            reset()
            return 0
        }

        let measuredAxis = axis / axisMagnitude
        if let previousAxis = stableAxis {
            guard simd_dot(previousAxis, measuredAxis) >= 0.75 else {
                accumulatedAngle = 0
                stableAxis = measuredAxis
                return 0
            }
            stableAxis = simd_normalize(previousAxis * 0.9 + measuredAxis * 0.1)
        } else {
            stableAxis = measuredAxis
        }

        accumulatedAngle += angleDelta
        guard accumulatedAngle >= 2 * .pi else { return 0 }

        let completed = Int(accumulatedAngle / (2 * .pi))
        accumulatedAngle.formTruncatingRemainder(dividingBy: 2 * .pi)
        return completed
    }
}

File: Android/app/src/main/java/com/azhegezhege/zhuzhiliao/physics/OrbitCounter.kt
package com.azhegezhege.zhuzhiliao.physics

import com.azhegezhege.zhuzhiliao.math.Vec3
import kotlin.math.PI

class OrbitCounter {
    private var accumulatedAngle = 0f
    private var stableAxis: Vec3? = null

    fun reset() {
        accumulatedAngle = 0f
        stableAxis = null
    }

    fun update(angleDelta: Float, axis: Vec3, isQualified: Boolean): Int {
        val axisMagnitude = axis.length
        if (!isQualified || angleDelta <= 0f || axisMagnitude <= Vec3.EPSILON) {
            reset()
            return 0
        }

        val measuredAxis = axis / axisMagnitude
        val previous = stableAxis
        if (previous != null) {
            if (previous.dot(measuredAxis) < 0.75f) {
                accumulatedAngle = 0f
                stableAxis = measuredAxis
                return 0
            }
            stableAxis = (previous * 0.9f + measuredAxis * 0.1f).normalized(measuredAxis)
        } else {
            stableAxis = measuredAxis
        }

        val fullTurn = (2.0 * PI).toFloat()
        accumulatedAngle += angleDelta
        if (accumulatedAngle < fullTurn) return 0
        val completed = (accumulatedAngle / fullTurn).toInt()
        accumulatedAngle %= fullTurn
        return completed
    }
}
